```
core.exception.AssertError@/home/runner/work/dmd/dmd/dmd/compiler/test/tools/dshell_prebuilt/dshell_prebuilt.d(204): last command failed
----------------
??:? _d_assert_msg [0x7fc3c8f8e3ac]
??:? [0x55cf5cadaedb]
??:? [0x55cf5cadb0e1]
??:? [0x55cf5cb38ee6]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll().__lambda2() [0x7fc3c8fcb31e]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x7fc3c8fcb1cd]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7fc3c8fcb2a6]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x7fc3c8fcb1cd]
??:? _d_run_main2 [0x7fc3c8fcb136]
??:? _d_run_main [0x7fc3c8fcaf1f]
??:? [0x55cf5cb38f87]
??:? __libc_start_main [0x7fc3c88ef0[82](https://github.com/dlang/dmd/actions/runs/3810503059/jobs/6482619242#step:18:83)]
??:? [0x55cf5cad9c3d]
Error: last command exited with code -11
==============================
>>> TARGET FAILED: dshell/dll_cxx.d
FAILED targets:
- dshell/dll_cxx.d
Error: Process completed with exit code 1.
```

@RazvanN7 Can we please fix the test suite so that the tests designed to find compiler bugs run *first* rather than *last*?

--- a/compiler/src/dmd/backend/cgelem.d
+++ b/compiler/src/dmd/backend/cgelem.d
@@ -762,92 +762,93 @@ private elem * elmemcmp(elem *e, goal_t goal)
 @trusted
 private elem * elmemset(elem *e, goal_t goal)
 {
+    //printf("elmemset()\n");
     elem_debug(e);
-    if (OPTIMIZER)
+
+    elem *ex = e.EV.E1;
+    if (ex.Eoper == OPnp_fp)
     {
-        elem *ex = e.EV.E1;
-        if (ex.Eoper == OPnp_fp)
-            eltonear(&ex);
-        else
-        {
-            // lvalue OPmemset (nbytes param value)
-            elem *enbytes = e.EV.E2.EV.E1;
-            elem *evalue = e.EV.E2.EV.E2;
-
-            version (MARS)
-            if (enbytes.Eoper == OPconst && evalue.Eoper == OPconst)
-            {
-                int nbytes = cast(int)el_tolong(enbytes);
-                targ_llong value = el_tolong(evalue);
-                elem *e1 = e.EV.E1;
-
-                if (e1.Eoper == OPcomma || OTassign(e1.Eoper))
-                    return cgel_lvalue(e);    // replace (e,v)op=e2 with e,(v op= e2)
-
-                tym_t tym;
-                switch (nbytes)
-                {
-                    case CHARSIZE:      tym = TYchar;   goto L1;
-                    case SHORTSIZE:     tym = TYshort;  goto L1;
-                    case LONGSIZE:      tym = TYlong;   goto L1;
-                    case LLONGSIZE:     if (_tysize[TYint] == 2)
-                                            goto default;
-                                        tym = TYllong;  goto L1;
-
-                    case 16:            if (REGSIZE == 8 && e1.Eoper == OPrelconst)
-                                            goto L1;
-                                        goto default;
-                    L1:
-                    {
-                        tym_t ety = e.Ety;
-                        memset(&value, value & 0xFF, value.sizeof);
-                        if (nbytes == 2 * REGSIZE && e1.Eoper == OPrelconst)
-                        {
-                            /* Rewrite as:
-                             * (*e1.0 = value),(*e1.4 = value),(e1)
-                             */
-                            tym = REGSIZE == 8 ? TYllong : TYint;
-
-                            auto e1a = el_copytree(e1);
-                            e1a.Eoper = OPvar;
-                            e1a.Ety = tym;
-                            auto ea = el_bin(OPeq,TYint,e1a,el_long(tym,value));
-
-                            auto e1b = el_copytree(e1);
-                            e1b.Eoper = OPvar;
-                            e1b.Ety = tym;
-                            e1b.EV.Voffset += REGSIZE;
-                            auto eb = el_bin(OPeq,tym,e1b,el_long(tym,value));
-
-                            e.EV.E1 = null;
-                            el_free(e);
-                            e = el_combine(el_combine(ea, eb), e1);
-                            e = optelem(e, GOALvalue);
-                            return e;
-                        }
-                        evalue.EV.Vullong = value;
-                        evalue.Ety = tym;
-                        e.Eoper = OPeq;
-                        e.Ety = (e.Ety & ~mTYbasic) | tym;
-                        if (tybasic(e1.Ety) == TYstruct)
-                            e1.Ety = tym;
-                        else
-                            e.EV.E1 = el_una(OPind, tym, e1);
-                        elem *tmp = el_same(&e.EV.E1);
-                        tmp = el_una(OPaddr, ety, tmp);
-                        e.EV.E2.Ety = tym;
-                        e.EV.E2 = el_selecte2(e.EV.E2);
-                        e = el_combine(e, tmp);
-                        e = optelem(e,GOALvalue);
-                    }
-                        break;
-
-                    default:
-                        break;
-                }
-            }
-        }
+        eltonear(&ex);
+        return e;
     }
+
+    // lvalue OPmemset (nelems param value)
+    elem *enelems = e.EV.E2.EV.E1;
+    elem *evalue = e.EV.E2.EV.E2;
+
+    if (!(enelems.Eoper == OPconst && evalue.Eoper == OPconst && REGSIZE >= 4))
+        return e;
+
+    elem *e1 = e.EV.E1;
+
+    if (e1.Eoper == OPcomma || OTassign(e1.Eoper))
+        return cgel_lvalue(e);    // replace ((e,v) memset e2) with (e,(v memset e2))
+
+    /* Attempt to replace OPmemset with a sequence of ordinary assignments,
+     * so cdmemset() will have fewer cases to deal with
+     */
+
+    const sz = tysize(evalue.Ety);
+    int nelems = cast(int)el_tolong(enelems);
+    ulong value = el_tolong(evalue);
+
+    if (sz * nelems > REGSIZE * 4)
+        return e;
+
+    switch (sz)
+    {
+        case 1:  value = 0x0101_0101_0101_0101 * (value & 0xFF);       break;
+        case 2:  value = 0x0001_0001_0001_0001 * (value & 0xFFFF);     break;
+        case 4:  value = 0x0000_0001_0000_0001 * (value & 0xFFFFFFFF); break;
+        case 8:  break;
+        default:
+            return e;
+    }
+
+    ulong valuexor = 0;
+    if (sz == 8 && REGSIZE == 4)
+        valuexor = (value ^ (value >> 32)) & 0xFFFF_FFFF;
+
+    elem* ey = null;
+    if (e1.Eoper != OPrelconst)
+    {
+        ey = e1;
+        e1 = el_same(&ey);
+    }
+    e.EV.E1 = null;             // so we can free e later
+
+    for (int offset = 0; offset < sz * nelems; )
+    {
+        int left = sz * nelems - offset;
+        if (left > REGSIZE)
+            left = REGSIZE;
+        tym_t tyv;
+        switch (left)
+        {
+            case 0: assert(0);
+            case 1: tyv = TYchar;  left = 1; break;
+            case 2:
+            case 3: tyv = TYshort; left = 2; break;
+            case 4:
+            case 5:
+            case 6:
+            case 7: tyv = TYlong;  left = 4; break;
+            default:
+            case 8: tyv = TYllong; left = 8; break;
+        }
+        auto e1a = el_copytree(e1);
+        assert(tybasic(e1a.Ety) != TYstruct);
+        e1a = el_bin(OPadd, TYnptr, e1a, el_long(TYsize_t, offset));
+        e1a = el_una(OPind, tyv, e1a);
+        auto ea = el_bin(OPeq, tyv, e1a, el_long(tyv, value));
+        if (sz * nelems - offset >= REGSIZE)
+            value ^= valuexor;          // flip between low and high 32 bits of 8 byte value
+        offset += left;
+        ey = el_combine(ey, ea);
+    }
+    ey = el_combine(ey, e1);
+    el_free(e);
+    e = optelem(ey,GOALvalue);
     return e;
 }
 

--- a/compiler/src/dmd/e2ir.d
+++ b/compiler/src/dmd/e2ir.d
@@ -6134,6 +6134,7 @@ StructDeclaration needsDtor(Type t)
  */
 elem *setArray(Expression exp, elem *eptr, elem *edim, Type tb, elem *evalue, IRState *irs, int op)
 {
+    //elem_print(evalue);
     assert(op == EXP.blit || op == EXP.assign || op == EXP.construct);
     const sz = cast(uint)tb.size();
     Type tb2 = tb;
@@ -6275,9 +6276,18 @@ Lagain:
             default:
                 assert(0);
         }
+        // do a cast to tym
         tym = tym | (evalue.Ety & ~mTYbasic);
-        evalue = addressElem(evalue, tb);
-        evalue = el_una(OPind, tym, evalue);
+        if (evalue.Eoper == OPconst)
+        {
+            evalue = el_copytree(evalue);
+            evalue.Ety = tym;
+        }
+        else
+        {
+            evalue = addressElem(evalue, tb);
+            evalue = el_una(OPind, tym, evalue);
+        }
     }
 
     evalue = useOPstrpar(evalue);

